KAFKA-16614:Disallow @ClusterTemplate("")

TaiJuWu opened the pull request:
As title.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensions.java:
```diff
@@ -123,8 +123,7 @@ private void processClusterTemplate(ExtensionContext context, ClusterTemplate an
         } else {
-            // Ensure we have at least one cluster config
-            generatedClusterConfigs.add(ClusterConfig.defaultClusterBuilder().build());
+            throw new IllegalStateException("Annotation value can't be empty string.");
```
Could we check `annot.value().isEmpty()` first? for example:
```java
if (annot.value().isEmpty()) throw exception;
```
With above change, we can simplify the if-else

chia7712 reviewed (commented):
@TaiJuWu thanks for this contribution. btw, is this check (https://github.com/apache/kafka/blob/81c222e9779c3339aa139ab930a74aba2c7c8685/core/src/test/java/kafka/test/junit/ClusterTestExtensions.java#L92) unnecessary?

TaiJuWu commented:
> @TaiJuWu thanks for this contribution. btw, is this check (
> 
> https://github.com/apache/kafka/blob/81c222e9779c3339aa139ab930a74aba2c7c8685/core/src/test/java/kafka/test/junit/ClusterTestExtensions.java#L92
> 
> ) unnecessary?

Yes, I believe it’s unnecessary. If we give default value (code on truncated branch), the size must be greater zero, we just need to select one.

Here we have two solutions, one is current version, the other is you look last time.

I don’t take any position, both are great for me.

chia7712 commented:
> I don’t take any position, both are great for me.

The current solution LGTM, and we can remove unnecessary condition check.

TaiJuWu commented:
> > I don’t take any position, both are great for me.
> 
> The current solution LGTM, and we can remove unnecessary condition check.

Thanks, the current version is the most simplified.

TaiJuWu committed:
Disallow @ClusterTemplate("")

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16614

chia7712 commented:
@TaiJuWu Could you add a new test file (`ClusterTestExtensionsUnitTest`) to verify the function `processClusterTemplate`?

TaiJuWu committed:
Add test file ClusterTestExtensionsUnitTest

chia7712 commented on core/src/test/java/kafka/test/ClusterTestExtensionsUnitTest.java:
```diff
@@ -0,0 +1,46 @@
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
+
+        Method method = ClusterTestExtensions.class.getDeclaredMethod("processClusterTemplate",
```
Maybe we can make `processClusterTemplate` be package-private, and then we just test that the method can throw exception when value is empty

TaiJuWu replied:
Done. Please review.

TaiJuWu committed:
Move ClusterTestExtensionsUnitTest into test/junit

1. Move Move ClusterTestExtensionsUnitTest into test/junit into core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java
2. Modify processClusterTemplate modifier to protected

chia7712 commented:
@TaiJuWu please fix the conflicts

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16614

TaiJuWu commented:
> @TaiJuWu please fix the conflicts

Done.

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensions.java:
```diff
@@ -118,16 +115,15 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
     }
 
-    private void processClusterTemplate(ExtensionContext context, ClusterTemplate annot,
+    protected void processClusterTemplate(ExtensionContext context, ClusterTemplate annot,
```
please remove `protected` to make it be package-private

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java:
```diff
@@ -0,0 +1,45 @@
+import static org.mockito.Mockito.when;
+
+public class ClusterTestExtensionsUnitTest {
+    @Test
```
please add `@SuppressWarnings("unchecked")`

TaiJuWu replied:
Done.

chia7712 reviewed (commented):
@TaiJuWu thanks for this fix. two minor comments left.

TaiJuWu committed:
Remove reflector.Method

TaiJuWu committed:
address comments

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensions.java:
```diff
@@ -91,9 +91,6 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
         if (clusterTemplateAnnot != null) {
             processClusterTemplate(context, clusterTemplateAnnot, generatedContexts::add);
-            if (generatedContexts.isEmpty()) {
-                throw new IllegalStateException("ClusterConfig generator method should provide at least one config");
```
My previous comment is wrong. This check is used to make sure there is at least one `ClusterConfig`, so we should not remove it. However, It seems to me this check should be moved to `processClusterTemplate` and we should add unit test for it.

TaiJuWu replied:
If we move the check to `processClusterTemplate`, `generatedContexts` also need to expose to `processClusterTemplate`, so I prefer keep this version and revert the change and try to add unit test for it.

chia7712 replied:
> try to add unit test for it.

nice. It is ok to file a jira to log it and we can merge this PR first.

chia7712 replied:
https://issues.apache.org/jira/browse/KAFKA-16652

@TaiJuWu Please feel free to take over it

TaiJuWu committed:
Restore testTemplate empry check

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensions.java:
```diff
@@ -118,16 +118,15 @@ public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContex
-            // Ensure we have at least one cluster config
-            generatedClusterConfigs.add(ClusterConfig.defaultBuilder().build());
+
+        if (annot.value().isEmpty()) {
```
How about `if (annot.value().trim().isEmpty()) throw new IllegalStateException("ClusterTemplate value can't be empty string");`

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java:
```diff
@@ -0,0 +1,45 @@
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
+
+        Assertions.assertThrows(IllegalStateException.class, () -> {
```
how about `Assertions.assertThrows(IllegalStateException.class, () -> ext.processClusterTemplate(context, annot, testInvocations));`?

chia7712 commented on core/src/test/java/kafka/test/junit/ClusterTestExtensionsUnitTest.java:
```diff
@@ -0,0 +1,45 @@
+        ExtensionContext context = mock(ExtensionContext.class);
+        Consumer<TestTemplateInvocationContext> testInvocations = mock(Consumer.class);
+        ClusterTemplate annot = mock(ClusterTemplate.class);
+        when(annot.value()).thenReturn("");
```
please add test for value "  "

TaiJuWu replied:
Done.

chia7712 reviewed (commented):
@TaiJuWu thanks for updated PR. it is almost there.

TaiJuWu committed:
Does not allow empty string or blank

chia7712 reviewed (approved):
LGTM
